src/oic/oic/provider.py

```diff
@@ -553,47 +544,118 @@ def end_session_endpoint(self, request="", cookie=None, **kwargs):
-            if not redirect_uri:
-                msg = "Post logout redirect URI verification failed!"
-                return error_response("%s", msg)
+        # 2 ways of find out client ID and user. Either through a cookie
+        # or using the id_token_hint
+        client_id = uid = None
+        sids = []
+        cd = None
 
-        authn, acr = self.pick_auth(esr)
+        if cookie:
+            cd = CookieDealer(srv=self)
+            _cval = cd.get_cookie_value(cookie, self.sso_cookie_name)
+            if _cval:
+                (value, _ts, typ) = _cval
+                if typ == 'sso':
+                    uid, client_id = value.split('][')
+                    try:
+                        sids = self.sdb.uid2sid[uid]
+                    except (KeyError, IndexError):
+                        return error_response('invalid_request', 'Mismatch uid')
 
-        sid = None
         if "id_token_hint" in esr:
-            id_token_hint = OpenIDRequest().from_jwt(esr["id_token_hint"],
-                                                     keyjar=self.keyjar,
-                                                     verify=True)
-            sub = id_token_hint["sub"]
+            id_token_hint = IdToken().from_jwt(esr["id_token_hint"],
+                                               keyjar=self.keyjar,
+                                               verify=True)
+            far_away = 86400*30  # 30 days
+
+            if client_id:
+                args = {'client_id': client_id}
+            else:
+                args = {}
+
             try:
-                # any sid will do, choose the first
-                sid = self.sdb.get_sids_by_sub(sub)[0]
-            except IndexError:
-                pass
-        else:
-            identity, _ts = authn.authenticated_as(cookie)
-            if identity:
-                uid = identity["uid"]
+                id_token_hint.verify(iss=self.baseurl, skew=far_away,
+                                     nonce_storage_time=far_away, **args)
+            except (VerificationError, NotForMe) as err:
+                logger.warning(
+                    'Verification error on id_token_hint: {}'.format(err))
+                return error_response('invalid_request', "Bad Id Token hint")
+
+            sub = id_token_hint["sub"]
+            if sids:
+                match = False
+                # verify that 'sub' are bound to 'user'
+                for sid in sids:
+                    if self.sdb[sid]['sub'] == sub:
+                        match = True
+                        break
+                if not match:
+                    return error_response('invalid_request', "Wrong user")
+
+            if not client_id:
+                if len(id_token_hint['aud']) == 1:
+                    client_id = id_token_hint['aud'][0]
+                else:
+                    client_id = id_token_hint['azp']
+
+            if not sids:
```

Comment:
This can be handled in `else` branch of `if` on line 585.

Reply:
OK